File: urbanairship-core/src/test/java/com/urbanairship/TestActivityMonitor.java
/* Copyright Airship and Contributors */

package com.urbanairship;

import android.app.Activity;
import androidx.annotation.NonNull;

import com.urbanairship.app.ActivityListener;
import com.urbanairship.app.ActivityMonitor;
import com.urbanairship.app.ApplicationListener;
import com.urbanairship.app.ForwardingActivityListener;
import com.urbanairship.app.ForwardingApplicationListener;

import java.util.ArrayList;
import java.util.List;

/**
 * Test Activity Monitor.
 */
public class TestActivityMonitor implements ActivityMonitor {

    private int startedActivities = 0;

    private ForwardingApplicationListener applicationListener = new ForwardingApplicationListener();
    private ForwardingActivityListener activityListener = new ForwardingActivityListener() {
        @Override
        public void onActivityResumed(Activity activity) {
            resumedActivities.add(activity);
            super.onActivityResumed(activity);
        }

        @Override
        public void onActivityPaused(Activity activity) {
            resumedActivities.remove(activity);
            super.onActivityPaused(activity);
        }

        @Override
        public void onActivityStarted(Activity activity) {
            startedActivities++;
            if (!isAppForegrounded) {
                isAppForegrounded = true;
                applicationListener.onForeground(System.currentTimeMillis());
            }

            super.onActivityStarted(activity);
        }

        @Override
        public void onActivityStopped(Activity activity) {
            if (startedActivities > 0) {
                startedActivities--;
            }

            if (startedActivities == 0 && isAppForegrounded) {
                isAppForegrounded = false;
                applicationListener.onBackground(System.currentTimeMillis());
            }

            super.onActivityStopped(activity);
        }
    };

    private boolean isAppForegrounded;
    private List<Activity> resumedActivities = new ArrayList<>();

    /**
     * Starts an activity.
     */
    public void startActivity() {
        Activity activity = new Activity();
        startActivity(activity);
    }

    /**
     * Stops an activity.
     */
    public void stopActivity() {
        Activity activity = new Activity();
        stopActivity(activity);
    }

    public void foreground(long timeStamp) {
        applicationListener.onForeground(timeStamp);
    }

    public void foreground() {
        isAppForegrounded = true;
        applicationListener.onForeground(0);
    }

    public void background() {
        isAppForegrounded = false;
        applicationListener.onBackground(0);
    }

    public void startActivity(Activity activity) {
        activityListener.onActivityStarted(activity);
    }

    public void resumeActivity(Activity activity) {
        activityListener.onActivityResumed(activity);
    }

    public void pauseActivity(Activity activity) {
        activityListener.onActivityPaused(activity);
    }

    public void stopActivity(Activity activity) {
        activityListener.onActivityStopped(activity);
    }

    @Override
    public void addActivityListener(@NonNull ActivityListener listener) {
        activityListener.addListener(listener);
    }

    @Override
    public void removeActivityListener(@NonNull ActivityListener listener) {
        activityListener.removeListener(listener);
    }

    @Override
    public void addApplicationListener(@NonNull ApplicationListener listener) {
        applicationListener.addListener(listener);
    }

    @Override
    public void removeApplicationListener(@NonNull ApplicationListener listener) {
        applicationListener.removeListener(listener);
    }

    @Override
    public boolean isAppForegrounded() {
        return isAppForegrounded;
    }

    @NonNull
    @Override
    public List<Activity> getResumedActivities() {
        return resumedActivities;
    }

    @NonNull
    @Override
    public List<Activity> getResumedActivities(@NonNull Predicate<Activity> filter) {
        List<Activity> activities = new ArrayList<>();

        for (Activity activity : resumedActivities) {
            if (filter.apply(activity)) {
                activities.add(activity);
            }
        }

        return activities;
    }

}

File: urbanairship-core/src/testFixtures/java/com/urbanairship/TestActivityMonitor.kt
/* Copyright Airship and Contributors */
package com.urbanairship

import android.app.Activity
import com.urbanairship.app.ActivityListener
import com.urbanairship.app.ActivityMonitor
import com.urbanairship.app.ApplicationListener
import com.urbanairship.app.ForwardingActivityListener
import com.urbanairship.app.ForwardingApplicationListener
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Test Activity Monitor.
 */
public class TestActivityMonitor : ActivityMonitor {

    private val _foregroundState = MutableStateFlow(false)
    override val foregroundState: StateFlow<Boolean> = _foregroundState.asStateFlow()

    private var startedActivities = 0
    private val applicationListener = ForwardingApplicationListener()
    private val activityListener: ForwardingActivityListener =
        object : ForwardingActivityListener() {
            override fun onActivityResumed(activity: Activity) {
                _resumedActivities.add(activity)
                super.onActivityResumed(activity)
            }

            override fun onActivityPaused(activity: Activity) {
                _resumedActivities.remove(activity)
                super.onActivityPaused(activity)
            }

            override fun onActivityStarted(activity: Activity) {
                startedActivities++
                if (!isAppForegrounded) {
                    isAppForegrounded = true
                    _foregroundState.value = true
                    applicationListener.onForeground(System.currentTimeMillis())
                }
                super.onActivityStarted(activity)
            }

            override fun onActivityStopped(activity: Activity) {
                if (startedActivities > 0) {
                    startedActivities--
                }
                if (startedActivities == 0 && isAppForegrounded) {
                    isAppForegrounded = false
                    applicationListener.onBackground(System.currentTimeMillis())
                }
                super.onActivityStopped(activity)
            }
        }
    override var isAppForegrounded: Boolean = false
        private set

    private var _resumedActivities: MutableList<Activity> = mutableListOf()
    public override val resumedActivities: List<Activity>
        get() = _resumedActivities


    /**
     * Starts an activity.
     */
    public fun startActivity() {
        val activity = Activity()
        startActivity(activity)
    }

    /**
     * Stops an activity.
     */
    public fun stopActivity() {
        val activity = Activity()
        stopActivity(activity)
    }

    public fun foreground(timeStamp: Long) {
        applicationListener.onForeground(timeStamp)
    }

    public fun foreground() {
        isAppForegrounded = true
        _foregroundState.value = true
        applicationListener.onForeground(0)
    }

    public fun background() {
        isAppForegrounded = false
        _foregroundState.value = false
        applicationListener.onBackground(0)
    }

    public fun startActivity(activity: Activity) {
        activityListener.onActivityStarted(activity)
    }

    public fun resumeActivity(activity: Activity) {
        activityListener.onActivityResumed(activity)
    }

    public fun pauseActivity(activity: Activity) {
        activityListener.onActivityPaused(activity)
    }

    public fun stopActivity(activity: Activity) {
        activityListener.onActivityStopped(activity)
    }

    override fun addActivityListener(listener: ActivityListener) {
        activityListener.addListener(listener)
    }

    override fun removeActivityListener(listener: ActivityListener) {
        activityListener.removeListener(listener)
    }

    override fun addApplicationListener(listener: ApplicationListener) {
        applicationListener.addListener(listener)
    }

    override fun removeApplicationListener(listener: ApplicationListener) {
        applicationListener.removeListener(listener)
    }

    override fun getResumedActivities(filter: Predicate<Activity>?): List<Activity> {
        val activities: MutableList<Activity> = ArrayList()
        for (activity in resumedActivities) {
            if (filter?.apply(activity) != false) {
                activities.add(activity)
            }
        }
        return activities
    }
}
